Dynamic commands and improved registration

PizzaCrust opened the pull request:
This commit adds registration handlers for implementations to handle command registration at any time, than just startup. This commit also adds dynamic commands called ICommand. It also fixes calling the same command name multiple times.

PizzaCrust committed:
Dynamic commands and improved registration

This commit adds registration handlers for implementations to handle command registration at any time, than just startup. This commit also adds dynamic commands called ICommand. It also fixes calling the same command name multiple times.

phase commented:
What are the use of dynamic commands? I think having immutable ones would be safer.

PizzaCrust commented:
Dynamic commands allows for plugins like Skript, to register commands dynamically after they parse the scripting language.

minecrafter commented:
I suggest that interfaces not be used as they are not safe. Please use an immutable object instead.

PizzaCrust commented:
Hmm. Probably like this:

``` java
Fountain.getServer().getCommandManager().newCommand("test-command").invokeHandler(() -> println("meow").register()
```

phase commented:
Or an object. Objects are nice. We're using Java for a reason.

PizzaCrust commented:
How would you propose to use an object for this?

phase commented:
Not called `Command` but something like

```
public abstract class Command {

    public abstract String getName();

    public abstract void invoke(String[] args, CommandSender sender);

}

```

Which plugins can extend like

```
public class TestCommand extends Command {

    public String getName() { return "test"; }

    public void invoke(String args[], CommandSender sender) { System.out.println(Arrays.toString(args)); }

}

```

PizzaCrust committed:
Command Builder for Dynamic Commands

This commit adds a CommandBuilder to generate ICommands, with immutable values.

phase commented on src/main/java/org/fountainmc/api/command/CommandBuilder.java:
```diff
@@ -0,0 +1,101 @@
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class CommandBuilder {
```
A builder to build an interface is redundant when you can create one class that can do the same thing.

PizzaCrust commented:
What's wrong with using builders?

``` java
Fountain.getServer().getCommandManager()
   .newCommand("test")
   .invokeHandler((args, sender) -> {})
   .generate();
```

PizzaCrust commented:
Ok, removing builders.

PizzaCrust committed:
Remove Command Builder and replace ICommand

This commit removes the CommandBuilder, and replaces ICommand to a class called AbstractCommand.

phase commented on src/main/java/org/fountainmc/api/command/RegistryHandler.java:
```diff
@@ -0,0 +1,12 @@
+    void onRegister(AbstractCommand iCommand);
+
+    void onRegister(CommandManager.CommandHandler commandHandler);
+}
```
You need some more new lines.

```
public class H {

    public void method() {

    }

}

```

Read the contributing file and add the formatters.

PizzaCrust commented:
Replaced with an abstract class.

PizzaCrust committed:
Add the whitespaces to conform to the contribution guidelines

This commit adds lines above the #onRegister methods.

PizzaCrust commented:
Added the lines.

PizzaCrust committed:
Fix the new whitespacing in CommandManager and AbstractCommand

This commit fixes the lines in CommandManager and AbstractCommand

phase commented:
Merged in https://github.com/FountainMC/FountainAPI/commit/73b702e2e72ceb808a32c8351d3032ffa93a0247
